#!/usr/bin/env python
PACKAGE = "amr_imu_filters"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

filter_enum = gen.enum([
    gen.const("Complementary", str_t, "COMPLEMENTARY", "Complementary filter"),
    gen.const("EKF", str_t, "EKF", "Extended Kalman Filter")
], "Filter type")

# 基本配置参数
gen.add("filter_type", str_t, 0, "Type of filter to use", "EKF", edit_method=filter_enum)
gen.add("use_mag", bool_t, 0, "Use magnetometer data", False)
gen.add("expected_publish_freq", double_t, 0, "Expected publish frequency", 0.0, 0.0, 200.0)  # 0表示使用回调发布

# 数据验证参数
gen.add("max_acceleration", double_t, 0, "Maximum allowable acceleration magnitude (m/s^2)", 32.0, 10.0, 50.0)
gen.add("max_angular_vel", double_t, 0, "Maximum allowable angular velocity magnitude (rad/s)", 14.0, 5.0, 20.0)

# 互补滤波器参数
gen.add("alpha", double_t, 0, "Complementary filter coefficient", 0.96, 0.0, 1.0)

# EKF参数
gen.add("process_noise_gyro", double_t, 0, "Process noise for gyroscope", 0.0001, 0.0, 0.1)
gen.add("process_noise_accel", double_t, 0, "Process noise for accelerometer", 0.001, 0.0, 0.1)
gen.add("process_noise_bias", double_t, 0, "Process noise for gyro bias", 0.00001, 0.0, 0.01)
gen.add("measurement_noise_accel", double_t, 0, "Measurement noise for accelerometer", 0.1, 0.0, 1.0)
gen.add("measurement_noise_mag", double_t, 0, "Measurement noise for magnetometer", 0.01, 0.0, 1.0)

# 初始化参数
gen.add("static_samples", int_t, 0, "Number of samples for static initialization", 100, 10, 1000)
gen.add("static_threshold", double_t, 0, "Threshold for static detection", 0.005, 0.001, 0.1)

# 通信参数
gen.add("subscriber_queue_size", int_t, 0, "Size of subscriber message queue", 50, 1, 1000)
gen.add("num_threads", int_t, 0, "Number of processing threads", 2, 1, 8)

exit(gen.generate(PACKAGE, "imu_filter", "ImuFilter"))